lib/git-lfs-s3/services/ceph_presigner.rb

```diff
@@ -0,0 +1,24 @@
+require 'base64'
+require 'date'
+require 'digest/sha1'
+require 'openssl'
+require 'uri'
+
+module GitLfsS3
+  module CephPresignerService
+    extend self
+    extend AwsHelpers
+
+    def signed_url(obj)
```

Comment:
Ok there is some explaining to do.

This is an implementation of the old deprecated AWS v1 Signer that is required for Ceph S3 (radosgw). It's no longer easily available in the current version of aws-sdk v2 (or even v1). This is because this signing method has been deprecated for at least a couple years and no longer works with AWS. I attempted to use v4 and v2 signatures without success. The [v4 signature](http://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html) is the current default for most AWS regions and it's completely ignored by radosgw. The [v2 signature](http://docs.aws.amazon.com/general/latest/gr/signature-version-2.html) builds a canonical request  (the string that is signed) that includes headers, content-length, etc. Radosgw does not include headers or content-length in it's canonical request. Eventually I looked at the Ceph S3 examples and radosgw source code for presigning and implemented code that would follow what they were doing instead of the existing S3 standards. That finally worked.

Re the name: I think it may also make sense to drop the Pre and call it CephSignerService. The reasoning for the pre is that it pre-signs a URL that is then returned and used out-of-band by an external service. Usually signers are used to validate an action that the library will manage and perform.